### ISSUE TITLE
Potential perf improvements

### ISSUE THREAD

**Issue by matklad** [near/near-memory-tracker#26]
0. Look at what APIs jemalloc itself has for introspection
1. Replace `pub static TID: RefCell<usize> = RefCell::new(0);` and the like with `pub static TID: Cell<usize> = Cell::new(0);`
2. Replace with const thread locals once it is stable https://github.com/rust-lang/rust/issues/84223
3. Replace with `thread::current().id().as_u64()` once that is stable
4. In general, minimize usage of thread-locals, they are quite slow in today's Rust
5. Spread out `MEM_SIZE` array more to avoid false sharing. Threads get sequential IDs, so may hammer the same cache line during allocation
6. Replace SeqCst with relaxed, I think that should be enough for our use case.
7. Remove `TOTAL_MEMORY_USAGE` as it's the sum of per-thread usages. 
7. `rand::thread_rng().gen_range(0, 100)` feels like something optimizable: thread_rng is crypto secure I thing (or at least stroger than xorshift).
7. https://mirrors.edge.kernel.org/pub/linux/kernel/people/paulmck/perfbook/perfbook.html I think uses "counting memory usage by threads" as one of the running examples, might be worth reading (warning: deep rabbit hole).
8. https://blog.mozilla.org/nnethercote/category/memory-allocation/ might also be an interesting read

**Comment by pmnoxx** [comment]
@matklad Thanks. I added a benchmark, which shows performance degradation: https://github.com/near/near-memory-tracker/pull/27

**Comment by pmnoxx** [comment]
> 7\. Remove `TOTAL_MEMORY_USAGE` as it's the sum of per-thread usages.

@matklad  I did some bench-marking, it saves round `0.4-0.6ns` from each run.
https://github.com/near/near-memory-tracker/pull/52

**Comment by pmnoxx** [comment]
> 8\. `rand::thread_rng().gen_range(0, 100)` feels like something optimizable: thread_rng is crypto secure I thing (or at least stroger than xorshift).

Yes, that's a big issue. I reduced time from `38ns` to `12ns`, by removing usage of `rand`. https://github.com/near/near-memory-tracker/pull/53

**Comment by pmnoxx** [comment]
> 6\. Replace SeqCst with relaxed, I think that should be enough for our use case.

I measured, there is no performance  between using `::Relaxed` vs `::SeqCst`.
See https://stackoverflow.com/questions/53805142/why-memory-order-relaxed-performance-is-the-same-as-memory-order-seq-cst/53805377
